### ISSUE TITLE
Solution to run with batch size larger than 1

### ISSUE THREAD

**Issue by dongliangcao** [nmwsharp/diffusion-net#10]
Dear author,

Thanks to your great work and code release for DiffusionNet. When I am working with DiffusionNet, I am wondering if there is a solution to train DiffusionNet with batch size larger than 1, even though the shape data might have different number of vertices.

Best regards, 
Dongliang

**Comment by nmwsharp** [comment]
Hi Dongliang,

Glad to hear you're interested!

There isn't currently any code built-in to this repo for training on irregular-sized batches.

There are two main strategies we could follow here:
- (A) Pad out the shorter entries with 0s or "fake" vertices so that all data has the same size. We might also want to incorporate "masks" of real/fake vertices so that some operations can ignore them (although I think the DiffusionNet code will mostly do the right thing if you just set the area to 0 for fake vertices). This is pretty easy, although it potentially wastes computation: the cost of the batch will be B * [biggest batch entry].
- (B) Concatenate all entries in to one giant combined entry, then process with batch size 1. This is the strategy that e.g. Pytorch Geometric uses. This mostly works pretty well, except some operations/losses need to be modified take these combined data as input.

If we were going to add something to this repo, I'd probably lean towards option (A). We could add some utilities to make it easy to pad out irregular inputs, and double-check that all operations behave sanely with respect to padding.

Let me know if that sounds reasonable! I probably wouldn't get to implementing anything until Feb at the earliest, I'm pretty busy at the moment.
